text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,239 @@
+
+# Motivation
+[motivation]: #motivation
+
+We have a number of situations where one crate defines an interface, and a different crate implements that interface with a canonical singleton:
+
+ - [`core::alloc::GlobalAlloc`](https://doc.rust-lang.org/nightly/core/alloc/trait.GlobalAlloc.html), chosen with [`#[global_allocator]`](https://doc.rust-lang.org/1.23.0/unstable-book/language-features/global-allocator.html)
+ - `panic_fmt` chosen with [`#[panic_implementation]`](https://github.com/rust-lang/rfcs/blob/master/text/2070-panic-implementation.md)
+ - The OOM hook, modified with [`std::alloc::{set,take}_alloc_error_hook`](https://doc.rust-lang.org/nightly/std/alloc/fn.set_alloc_error_hook.html)
+ - [`std::collections::hash_map::RandomState`](https://doc.rust-lang.org/std/collections/hash_map/struct.RandomState.html), if https://github.com/rust-lang/rust/pull/51846 is merged, the `hashmap_random_keys` lang item
+ - [`log::Log`](https://docs.rs/log/0.4.3/log/trait.Log.html) set with [`log::set_logger`](https://docs.rs/log/0.4.3/log/fn.set_logger.html)
+
+Each of these is an instance of the same general pattern.
+But the solutions are all ad-hoc and distinct, burdening the user of Rust and rustc with extra work remembering/implementing, and preventing more rapid prototyping.
+
+They also incur a run-time cost due to dynamism and indirection, which can lead to initialization bugs or bloat in space-constrained environments.
+In the annotation case, there's essentially an extra `extern { fn special_name(..); }` whose definition the annotation generates.
+This isn't easily inlined outside of LTO, and even then would prohibit rustc's own optimizations going into affect.
+The `set`-method based ones involve mutating a `static mut` or equivalent with a function or trait object, and thus can basically never be inlined away.
+So there's the overhead of the initialization, and then one or two memory dereferences to get the implementation function's actual address.
+The potential bugs are due to not `set`ing before the resource is needed, a manual task because there's static way to prevent accessing the resource while it isn't set.
+
+The `extern existential type` feature just covers the deferred definition of a type, and not the singleton itself, but that is actually enough. For example, with global allocation:
+
+```rust
+// In `alloc`
+
+pub extern existential type Heap: Copy + Alloc + Default + Send + Sync;
+
+struct Box<T, A: Alloc = Heap>;
+
+impl Box<T, A: Alloc> {
+    fn new_in(a: A) { .. }
+}
+
+impl Box<T, A: Alloc + Default = Heap> {
+    fn new() { Self::new_in(Default::default())    }
+}
+```
+
+```rust
+// In `jemalloc`
+
+#[deriving(Default, Copy)]
+struct JemallocHeap;
+
+impl Alloc for JemallocHeap {
+    fn alloc(stuff: Stuff) -> Thing {
+        ...
+    }
+}
+
+extern existential type alloc::Heap = JemallocHeap;
```

Comment:
nit: This declaration wouldn't live in the jemalloc crate, but rather the downstream consumer. We want to allow people to pull in jemalloc without using it as the global allocator. (Imagine you want to wrap it in a layer of tracking or whatever)

Reply:
Yes good point thanks @sfackler. I'll amend that example to show the 3rd crate.